crates/next-error-code-swc-plugin/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+
+        // Maintain string_occurrences
+        *self
+            .string_occurrences
+            .entry(error_message.clone())
+            .or_insert(0) += 1;
+
+        // Generate the hash of object { file_path, error_message, occurrence_count }
+        let mut hasher = DefaultHasher::new();
+        serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string()
+        .hash(&mut hasher);
+        let hash = format!("{:08x}", hasher.finish());
+
+        // Concatenate the commit hash and the hash
+        let error_code = format!("E{}{}", self.commit_hash, hash);
+
+        // Save to JSON file
+        let error_metadata = serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string();
+        let mut retries = 3;
+
+        // Do not do FS operations in unit tests
+        if cfg!(test) {
+            return error_code;
+        }
+
+        loop {
+            if self.mode == "check" {
+                let error_code_path = format!("cwd/error_codes/{}.json", hash);
+                if !std::path::Path::new(&error_code_path).exists() {
+                    panic!(
+                        "ERROR: File {} does not exist.\n\nREQUIRED ACTION:\n1. Run `pnpm \
+                         build`\n2. Commit all file changes from \
+                         /packages/next/error-codes\n\nThis is required to maintain error code \
+                         consistency.",
+                        format!("/packages/next/error_codes/{}.json", hash)
+                    );
+                }
+                return error_code;
+            }
+
+            if self.mode != "generate" {
+                panic!("Mode must be 'generate', got '{}'", self.mode);
+            }
+
+            fs::create_dir_all("cwd/error_codes").unwrap_or_else(|e| {
+                panic!("Failed to create error_codes directory: {}", e);
+            });
+
+            match fs::write(format!("cwd/error_codes/{}.json", hash), &error_metadata) {
+                Ok(_) => break,
+                Err(e) => {
+                    retries -= 1;
+                    if retries == 0 {
+                        panic!("Failed to write error metadata after 3 retries: {}", e);
+                    }
+                }
+            }
+        }
+
+        error_code
+    }
+}
+
+impl VisitMut for TransformVisitor {
+    fn visit_mut_expr(&mut self, expr: &mut Expr) {
+        expr.visit_mut_children_with(self);
+
+        let mut code: Option<String> = None;
+        let mut new_error_expr: Option<&NewExpr> = None;
+
+        if let Expr::New(new_expr) = expr {
+            if let Expr::Ident(ident) = &*new_expr.callee {
+                if ident.sym.to_string() == "Error" {
```

Comment:
These are all the Error subclasses we need to check:
````
const errorConstructors = [
  // built-ints
  'AggregateError',
  'Error',
  'EvalError',
  'RangeError',
  'ReferenceError',
  'SyntaxError',
  'TypeError',
  'URIError',
  // custom
  'ApiError',
  'BailoutToCSRError',
  'BubbledError',
  'CanaryOnlyError',
  'Cancel',
  'CompileError',
  'CssSyntaxError',
  'DecodeError',
  'DynamicServerError',
  'ExportError',
  'FatalError',
  'ImageError',
  'InvariantError',
  'ModuleBuildError',
  'NestedMiddlewareError',
  'NoFallbackError',
  'NoSuchDeclarationError',
  'PageSignatureError',
  'PostCSSSyntaxError',
  'ReadonlyHeadersError',
  'ReadonlyRequestCookiesError',
  'ReadonlyRequestCookiesError',
  'ReadonlyURLSearchParamsError',
  'ResponseAborted',
  'SerializableError',
  'StaticGenBailoutError',
  'TimeoutError',
  'Warning',
]
```
found by checking which classes actually subclass `Error` via [`[collect-error-constructors.mjs](https://gist.github.com/eps1lon/6cce3059dfa061f2a7dc28305fdaddae#file-collect-error-constructors-mjs)`](https://gist.github.com/eps1lon/6cce3059dfa061f2a7dc28305fdaddae#file-collect-error-constructors-mjs)